kore/src/Kore/Internal/SideCondition.hs

```diff
@@ -908,3 +940,115 @@ generateNormalizedAcs internalAc =
 isDefinedInternal :: TermLike variable -> Bool
 isDefinedInternal =
     Attribute.isDefined . TermLike.termDefined . TermLike.extractAttributes
+
+fromSimplifiedFunctions ::
+    InternalVariable variable =>
+    HashSet (Application Symbol (TermLike variable)) ->
+    SideCondition variable
+fromSimplifiedFunctions simplifiedFunctions =
+    top{simplifiedFunctions}
+
+{- | Stores all non-constructor function symbols appearing in a term.
+ Inside a rewrite step, this information is used to avoid trying to
+ reevaluate functions which could not be evaluated during the 'Simplify'
+ stage of execution.
+
+ See 'isSimplifiedFunction'.
+-}
+cacheSimplifiedFunctions ::
+    forall variable.
+    InternalVariable variable =>
+    TermLike variable ->
+    SideCondition variable
+cacheSimplifiedFunctions =
+    fromSimplifiedFunctions
+        . extractSimplifiedFunctions
+  where
+    extractSimplifiedFunctions ::
+        TermLike variable ->
+        HashSet (Application Symbol (TermLike variable))
+    extractSimplifiedFunctions (Recursive.project -> _ :< termF) =
+        case termF of
+            TermLike.ApplySymbolF symbolApp ->
+                let symbol = TermLike.applicationSymbolOrAlias symbolApp
+                    children = TermLike.applicationChildren symbolApp
+                 in if isFunction symbol && not (isConstructor symbol)
```

Comment:
Marked simplified in what way, you mean the `Simplified` attribute?

Reply:
Yeah

Reply:
No, the check based on this function is stronger than the `Simplified` attribute. We call this after we know that a function has been fully simplified during a rewrite step, to avoid trying to resimplify during that step. The plan is to eventually remove the `Simplified` attribute.